src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,200 @@
+                currentNode = currentNode.prev;
+            }
+            return currentNode;
+        }
+    }
+
+    private void linkLast(T element) {
+        final Node<T> newNode = new Node<>(tail, element, null);
+        if (head == null) {
+            head = newNode;
+        } else {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        size++;
+    }
+
+    private boolean isValidIndex(int index) {
+        return index >= 0 && index < size;
+    }
+
+    private void linkBefore(T value, Node<T> node) {
+        final Node<T> newNode = new Node<>(node.prev, value, node);
+        if (node.prev == null) {
+            head = newNode;
+        } else {
+            node.prev.next = newNode;
+        }
+        node.prev = newNode;
+        size++;
+    }
+
+    private T unlinkFirst(Node<T> headNode) {
+        final T element = headNode.item;
+        final Node<T> next = headNode.next;
+        headNode.item = null;
+        headNode.next = null;
+        head = next;
+        if (next == null) {
+            tail = null;
+        } else {
+            next.prev = null;
+        }
+        size--;
+        return element;
+    }
+
+    private T unlinkLast(Node<T> tailNode) {
+        final T element = tailNode.item;
+        final Node<T> prev = tailNode.prev;
+        tailNode.item = null;
+        tailNode.prev = null;
+        tail = prev;
+        if (prev == null) {
+            head = null;
+        } else {
+            prev.next = null;
+        }
+        size--;
+        return element;
+    }
+
+    private T unlink(Node<T> node) {
+        final T element = node.item;
+        if (node.prev == null) {
+            head = node.next;
+        } else {
+            node.prev.next = node.next;
+        }
+        if (node.next == null) {
+            tail = node.prev;
+        } else {
+            node.next.prev = node.prev;
+        }
+        node.prev = null;
+        node.next = null;
+        node.item = null;
+        size--;
+        return element;
+    }
+
     @Override
     public void add(T value) {
+        linkLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        if (!isValidIndex(index) && index != size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        if (index == size) {
+            linkLast(value);
+        } else {
+            linkBefore(value, node(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list.size() == 0) {
+            throw new RuntimeException("You can't add empty collection.");
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (!isValidIndex(index)) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
```

Comment:
it's will be better if you throw your exception inside the isValidIndex(index) method, this will allow us to remove this if check and we don't have to throw exception here as well, do it for all methods where you use isValidIndex(index)